pkg/v2/tkr/webhook/cluster/tkr-resolver/cluster/cluster.go

```diff
@@ -6,16 +6,353 @@ package cluster
+				getMap(&cluster.Spec.Topology.Workers.MachineDeployments[i].Metadata.Labels),
+				getMap(&cluster.Spec.Topology.Workers.MachineDeployments[i].Metadata.Annotations))
+		}
+	}
+}
+
+func setMetadataFromOSImageResult(osImageResult *data.OSImageResult, ls, annots map[string]string) {
+	for osImageName, osImage := range osImageResult.OSImagesByTKR[osImageResult.TKRName] { // only one such OSImage
+		ls[runv1.LabelOSImage] = osImageName
+
+		ls[runv1.LabelOSType] = osImage.Spec.OS.Type
+		ls[runv1.LabelOSName] = osImage.Spec.OS.Name
+		ls[runv1.LabelOSVersion] = osImage.Spec.OS.Version
+		ls[runv1.LabelOSArch] = osImage.Spec.OS.Arch
+
+		bytes, _ := yaml.Marshal(osImage.Spec.Image.Ref) // error always nil: data from the API Server is safe
+		annots[runv1.AnnotationOSImageRef] = string(bytes)
+
+		ls[runv1.LabelImageType] = osImage.Spec.Image.Type
+		osimage.SetRefLabels(ls, osImage.Spec.Image.Type, osImage.Spec.Image.Ref)
+	}
+}
+
+// getMap returns the map (creates it first if the map is nil). mp has to be a pointer to the variable holding the map,
+// so that we could save the newly created map.
+// Pre-reqs: mp != nil
+func getMap(mp *map[string]string) map[string]string { // nolint:gocritic // suppress warning: ptrToRefParam: consider `mp' to be of non-pointer type (gocritic)
+	if *mp == nil {
+		*mp = map[string]string{}
+	}
+	return *mp
+}
+
+// getClusterClass gets ClusterClass for the cluster. Returns nil if the cluster is not classy.
+// Pre-reqs: cluster != nil
+func (cw *Webhook) getClusterClass(ctx context.Context, cluster *clusterv1.Cluster) (*clusterv1.ClusterClass, error) {
+	if cluster.Spec.Topology == nil {
+		return nil, nil
+	}
+	clusterClass := &clusterv1.ClusterClass{}
+	if err := cw.Client.Get(ctx, client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Spec.Topology.Class,
+	}, clusterClass); err != nil {
+		return nil, err
+	}
+	return clusterClass, nil
+}
+
+// success constructs PatchResponse from the mutated cluster.
+func success(req *admission.Request, cluster *clusterv1.Cluster) admission.Response {
+	marshaledCluster, err := json.Marshal(cluster)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledCluster)
+}
+
+func (cw *Webhook) getTKR(cluster *clusterv1.Cluster) *runv1.TanzuKubernetesRelease {
+	if cluster.Labels == nil {
+		return nil
+	}
+	tkrName, exists := cluster.Labels[runv1.LabelTKR]
+	if !exists {
+		return nil
+	}
+	return cw.TKRResolver.Get(tkrName, &runv1.TanzuKubernetesRelease{}).(*runv1.TanzuKubernetesRelease)
+}
+
+const VarTKRKubernetesSpec = "TKR_KUBERNETES_SPEC"
+
+func adjustClusterKubernetesSpec(tkr *runv1.TanzuKubernetesRelease, clusterClass *clusterv1.ClusterClass, cluster *clusterv1.Cluster) {
+	cluster.Spec.Topology.Version = tkr.Spec.Kubernetes.Version
+
+	if topology.ClusterClassVariable(clusterClass, VarTKRKubernetesSpec) == nil {
+		return
```

Comment:
Should we log that we are essentially skipping setting of the TKR_KUBERNETES_SPEC variable?

Reply:
Not suggesting we set, the suggestion is that we log that we are _not_ setting.